packages/gatsby/lib/query-runner/file-parser.js

```diff
@@ -0,0 +1,133 @@
+// @flow
+const fs = require(`fs`)
+const crypto = require(`crypto`)
+
+// Traverse is a es6 module...
+import traverse from 'babel-traverse'
+import * as types from 'babel-types'
+const babylon = require(`babylon`)
+const Bluebird = require(`bluebird`)
+
+const apiRunnerNode = require(`../utils/api-runner-node`)
+const { getGraphQLTag } = require(`../utils/babel-plugin-extract-graphql`)
+
+import type { DocumentNode, DefinitionNode } from 'graphql'
+
+const readFileAsync = Bluebird.promisify(fs.readFile)
+
+function getAssignedIdenifier(path) {
+  let property = path.parentPath
+  while (property) {
+    if (types.isVariableDeclarator(property)) return property.node.id.name
+    property = property.parentPath
+  }
+}
+
+async function parseToAst(filePath, fileStr) {
+  let ast
+
+  // Preprocess and attempt to parse source; return an AST if we can, log an
+  // error if we can't.
+  const transpiled = await apiRunnerNode(`preprocessSource`, {
+    filename: filePath,
+    contents: fileStr,
+  })
+
+  if (transpiled.length) {
+    for (const item of transpiled) {
+      try {
+        const tmp = babylon.parse(item, {
+          sourceType: `module`,
+          plugins: [`*`],
+        })
+        ast = tmp
+        break
+      } catch (e) {
+        console.info(e)
+        continue
+      }
+    }
+    if (ast === undefined) {
+      console.error(`Failed to parse preprocessed file ${filePath}`)
+    }
+  } else {
+    try {
+      ast = babylon.parse(fileStr, {
+        sourceType: `module`,
+        sourceFilename: true,
+        plugins: [`*`],
+      })
+    } catch (e) {
+      console.log(`Failed to parse ${filePath}`)
+      console.log(e)
+    }
+  }
+
+  return ast
+}
+
+async function findGraphQLTags(file, text): Promise<Array<DefinitionNode>> {
+  let ast = await parseToAst(file, text)
+  if (!ast) return []
+
+  let queries = []
+  traverse(ast, {
+    ExportNamedDeclaration(path, state) {
+      path.traverse({
+        TaggedTemplateExpression(innerPath) {
+          const gqlAst = getGraphQLTag(innerPath)
+          if (gqlAst) {
+            gqlAst.definitions.forEach(def => {
+              if (!def.name || !def.name.value) {
+                if (gqlAst.definitions.length === 1)
+                  def.name = { value: getAssignedIdenifier(innerPath) }
```

Comment:
if you don't name the fragment/query it'll use the identifier name that is exported. eg

```
export foo = graphql` query { foobar }`
```
will be named `foo`